### PR TITLE
control-service: data job execution statuses refinement

### PR DESCRIPTION
**Pull request by mivanov1988**
Currently the Job execution statuses are not well designed.
We have the following statuses:

    SUBMITTED
    RUNNING
    FINISHED
    FAILED
    CANCELLED
    SKIPPED

Where the FAILED represents two statuses: USER_ERROR and PLATFORM_ERROR.
One of the our internal clients wants to see the exact status
(e.g. USER ERROR rather than FAILED). Also the client wants to filter
the executions by USER_ERROR or PLATFORM_ERROR.

The FAILED status is split into USER_ERROR and PLATFORM_ERROR. Also the
FINISHED is renamed to SUCCEEDED.

Testing done: unit and integration tests pass

Signed-off-by: Miroslav Ivanov miroslavi@vmware.com

### CODE REVIEW

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/execution/JobExecutionResultManager.java [hunk 1]
```diff
@@ -35,130 +34,97 @@
...
-    * Updates job execution termination status based on job execution status.
-    * In case of missing termination status due to the missing K8S Pod
-    * it sets an appropriate termination status based on the execution status.
+    * Returns execution status based on K8S Job status.
     *
-    * @param terminationStatus termination status based on the K8S Pod termination status
-    * @param executionStatus execution status based on K8S Job status
-    * @param terminationStatusString termination status returned from K8S Pod (e.g. "Success", "User error", etc.)
     * @param jobTerminationReason condition reason as reported by K8s Job (e.g. "DeadlineExceeded", "BackoffLimitExceeded", etc.)
     * @param containerTerminationReason termination reason for pod container as reported by K8s Job (e.g., "OOMKilled", etc.)
```
Comment (tpalashki) [inline projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/execution/JobExecutionResultManager.java:117]: Comment should be:
> ...as returned by K8s pod container (e.g., "OOMKilled", etc.)

File: projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLExecutionsIT.java [hunk 2]
```diff
@@ -115,34 +116,6 @@ public void testExecutions_filterByStartTimeGte() throws Exception {
                   Matchers.not(Matchers.contains(dataJobExecution3.getId()))));
    }
 
-   @Test
```
Comment (tpalashki) [inline projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLExecutionsIT.java:118]: What's the reason to remove this test?

File: projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLJobExecutionsStatusCountIT.java [hunk 3]
```diff
@@ -102,10 +102,10 @@ public void testExecutionStatusCount_expectTwoSuccessfulTwoFailed(String jobName
...
       var expectedEndTimeSmaller = OffsetDateTime.now().minusDays(1);
 
-      addJobExecution(expectedEndTimeLarger, "testId", ExecutionStatus.FINISHED, jobName);
-      addJobExecution(expectedEndTimeSmaller, "testId2", ExecutionStatus.FINISHED, jobName);
-      addJobExecution(expectedEndTimeLarger, "testI3", ExecutionStatus.FAILED, jobName);
-      addJobExecution(expectedEndTimeSmaller, "testId4", ExecutionStatus.FAILED, jobName);
+      addJobExecution(expectedEndTimeLarger, "testId", ExecutionStatus.SUCCEEDED, jobName);
+      addJobExecution(expectedEndTimeSmaller, "testId2", ExecutionStatus.SUCCEEDED, jobName);
+      addJobExecution(expectedEndTimeLarger, "testI3", ExecutionStatus.PLATFORM_ERROR, jobName);
+      addJobExecution(expectedEndTimeSmaller, "testId4", ExecutionStatus.PLATFORM_ERROR, jobName);
```
Comment (tpalashki) [inline projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLJobExecutionsStatusCountIT.java:108]: Maybe introduce somewhere around here an execution with a USER_ERROR to test this use case.

File: projects/control-service/projects/pipelines_control_service/src/test/java/com/vmware/taurus/service/JobExecutionRepositoryIT.java [hunk 4]
```diff
@@ -166,15 +166,15 @@ void testFindFirstByDataJobNameOrderByStartTimeDesc_shouldReturnTheLastExecution
...
+              "execution1", dataJob, ExecutionStatus.SUCCEEDED, "Success",
               OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC),
               OffsetDateTime.of(2000, 1, 1, 1, 0, 0, 0, ZoneOffset.UTC));
       RepositoryUtil.createDataJobExecution(jobExecutionRepository,
               "execution2", dataJob, ExecutionStatus.RUNNING, null,
               OffsetDateTime.of(2000, 1, 1, 4, 0, 0, 0, ZoneOffset.UTC),
               null);
       RepositoryUtil.createDataJobExecution(jobExecutionRepository,
-              "execution3", dataJob, ExecutionStatus.FAILED, "User error",
+              "execution3", dataJob, ExecutionStatus.PLATFORM_ERROR, "User error",
```
Comment (tpalashki) [inline projects/control-service/projects/pipelines_control_service/src/test/java/com/vmware/taurus/service/JobExecutionRepositoryIT.java:177]: The message is a bit misleading: "User error" vs PLATFORM_ERROR status.

File: projects/control-service/projects/pipelines_control_service/src/test/java/com/vmware/taurus/service/monitoring/DataJobMonitorTest.java [hunk 5]
```diff
@@ -561,15 +579,15 @@ void testRecordJobExecutionStatus_withAnOlderExecution_shouldNotUpdateLastExecut
         Optional<DataJob> actualJob = jobsRepository.findById(jobExecution.getJobName());
         Assertions.assertFalse(actualJob.isEmpty());
         // The last execution status should not have changed from FINISHED to FAILED because the execution is not recent
```
Comment (tpalashki) [inline projects/control-service/projects/pipelines_control_service/src/test/java/com/vmware/taurus/service/monitoring/DataJobMonitorTest.java:581]: Update this comment.

**Review by tpalashki — APPROVED** [review]
LGTM. Thanks for making those changes, things look much clearer now.
Update the CHANGELOG as we introduce changes to the API.

File: projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLExecutionsIT.java (hunk 2 above)
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLExecutionsIT.java:118]: I haven't merged well

**Comment by mivanov1988** [comment]
@tpalashki Thank you for the review.

File: projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/model/converter/ExecutionStatusConverter.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+   }
+
+   @Override
+   public ExecutionStatus convertToEntityAttribute(Integer dbValue) {
+      if (dbValue == null) {
+         return null;
+      }
+
+      // temporary solution for backward compatibility.
+      if (dbValue == 3) {
```
[the remark this answers is not in the document]
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/model/converter/ExecutionStatusConverter.java:29]: I prepared the following queries which I am planning to execution right after the deployment:
```
update data_job_execution set status = 6 where message = 'User error' and status = 3;
update data_job_execution set status = 7 where message = 'Platform error' and status = 3;
update data_job_execution set status = 2 where message = 'Success' and status = 3;
```

File: projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLJobExecutionsStatusCountIT.java (hunk 3 above)
Comment (mivanov1988) [reply projects/control-service/projects/pipelines_control_service/src/integration-test/java/com/vmware/taurus/graphql/it/GraphQLJobExecutionsStatusCountIT.java:108]: I added the PLATFORM_ERROR only to test that the API will work with status that contains underscore.

**Pull request merged**